Mapper visualisation refactor: fix bugs, add summary statistics in hovertext, improve opacity, remove matplotlib dependency, add node_scale kwarg, add clone_pipeline kwarg to interactive plots,restructure/rename plotly_kwargs, improve code

ulupo opened the pull request:
**Reference issues/PRs**
Fixes #399, #403 and #404 (thanks @miltminz for reporting these). It is an alternative to #405 by @MikeG112, developed in parallel to that solution. It also contains an implementation of a part of #382.

**Types of changes**
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
This is a thorough refactoring of `gtda.mapper`, It achieves the following:
- It makes the code clearer and shorter, and removes several redundancies.
- It fixes issues #399 and #403 concerning node coloring. Concerning #399, there was a simple bug in the code responsible to scaling between 0 and 1. However, *[0, 1]-scaling has been completely removed* from the code. This is because `plotly` colorscales are capable of handling arbitrary floats. Concerning #403, the color bar now *always shows the true range*, without rescaling to [0, 1].  
- It adds the node summary statistics data to the node hovertext, rounding it to a user-specified number of significant figures controlled by a *new parameter* `n_sig_figs`. This addresses feature request #404.
- It *renames* `plotly_kwargs` to `plotly_params`, slightly *changes the specifications* for it, and slightly changes the logic for updating plotly figures according to this parameter.
- It *removes* the need for `matplotlib`'s `rgb2hex` function and hence the need for the `matplotlib` requirement in `giotto-tda` altogether. This is achieved via a new helper function `_get_colors_for_vals` performing colorscale-based interpolations.
- It adds a `node_scale` kwarg to the static and interactive plotting functions and makes `set_node_sizeref` private.
- It adds a `clone_pipeline` kwarg to `plot_interactive_mapper_graph`, overlapping with #382.
- It improves the overall look of the Mapper graph, particularly by increasing the opacity of node colors so that edges do not hide them.

**Checklist**
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

Commit:
First go at mapper refactoring

Commit:
Remove Nones from buttons

Commit:
Use real colorscale

Commit:
Merge branch 'master' into mapper/fix_mapper_colors

Commit:
Add test-output.xml to .gitignore

Commit:
Replace "text" key in plot_options["node_trace"] with "hovertext", fix tests

Commit:
Change variable name to comply with E741

Commit:
Merge branch 'master' into mapper/fix_mapper_colors

wreise reviewed (commented):
LGTM (i tested it briefly from a user perspective), just two things:
1. The plotly arguments were useful, but maybe they need to be compromised,
2. Is it possible to format  the strings in summary statstics to display only a few digits?

Commit:
Implement significant figure rounding in hovertext, add n_sig_figs kwargs

Commit:
Improve width and opacity of nodes and edges

Commit:
Extend support of summary statistics in hovertext for columns in dropdown

- Minor code improvements
- Revert relative orders of node_trace and edge_trace
- Docstring improvements
- Linting

Commit:
Attempt removing rgb2hex

Commit:
Remove cmin and cmax, rename variables, change return signature of calculate_graph_data

Commit:
Fix tests

Commit:
Linting

Commit:
Fix lack of dropdowns in interactive graph

Commit:
Remove unnecessary dimensionality check in layout_dim

Commit:
Fix typo

Commit:
Fix test

Commit:
Rename plotly_kwargs to plotly_params

Commit:
Remove colorscale from output of _calculate_graph_data, improve node_pos variable, rename some variables

Commit:
Fix test typo

Commit:
Remove unnecessary np.asarray so test can be reverted

Commit:
Fix test now that node_trace x and y data is numpy array

Commit:
Reinstate plotly_params functionality

Commit:
Remove matplotlib installation requirement

Commit:
Fix mapper notebook

Commit:
Add small comment on use of matplotlib in voids_on_the_plane

Commit:
Fix linting

Commit:
Place matplotlib in examples requirement in setup.py

Commit:
Remove bgcolor from color buttons as it seems to have no effect, reinstate Nones even if it seems to make no difference

Commit:
Fix issue with values to unpack in on_parameter_change

Commit:
Improve docstring of make_mapper_pipeline

lewtun commented on gtda/mapper/pipeline.py:
```diff
@@ -216,14 +216,13 @@ def make_mapper_pipeline(scaler=None,
-        pipeline. Caching the transformers is advantageous when fitting is
-        time consuming.
+        Used to cache the fitted transformers which make up the pipeline. This
+        is advantageous when fitting of early steps is time consuming and only
```
i would write "when the fitting ..."

ulupo replied:
Yep, indeed. Excessive copy/pasting from `scikit-learn` is not a good excuse :).

lewtun reviewed (approved):
LGTM

ulupo committed:
Improve grammar

Commit:
Improve examples for make_mapper_pipeline and create one for plot_static_mapper_graph

Commit:
Hide set_node_sizeref, add node_scale kwarg, add clone_pipeline kwarg to plot_interactive_mapper_graph

- Update docstrings
- Update quickstart notebook

Commit:
Add helper function for colorscale-based interpolations, improve use of plotly_params

Static and interactive plots display the expect hoverlabel colors in 3D, or white if things go wrong. Thanks to @lewtun for pointing out that this was needed.

Commit:
Improve docstrings

lewtun commented on gtda/mapper/pipeline.py:
```diff
@@ -258,16 +258,17 @@ def make_mapper_pipeline(scaler=None,
-              f'Data points: {X[node_elements[0]]}')
+    >>> node_id = mapper_graph['node_metadata']['node_id']
+    >>> node_elements = mapper_graph['node_metadata']['node_elements']
+    >>> print(f"Node Id: {node_id[0]}, Node elements: {node_elements[0]}, "
```
Should this be "Node ID" for consistency elsewhere in our docs?

lewtun reviewed (approved):
Nice work on making the 3D hoverlabel colours work! Overall the PR LGTM and can be merged!

ulupo committed:
Name fix after @lewtun's comment

Commit:
Hide visualization module in mapper/utils
